crates/rustc_codegen_spirv/src/linker/structurizer.rs

```diff
@@ -341,20 +560,39 @@ fn get_possible_merge_positions(
             }
 
             // Make sure we are not continuing after a merge.
-            if block_is_parent_of(loop_info.header_id, start, blocks) && front == loop_info.merge_id
-            {
+            if block_is_parent_of(loop_info.header_id, a, blocks) && front == loop_info.merge_id {
                 new_edges.clear();
             }
         }
 
-        // We found a possible merge position, make sure it isn't a merge of a loop because in that case we want to use break logic.
-        if new_edges.len() == 1 && !cf_info.id_is_loops_merge(new_edges[0]) {
-            retval.push(find_block_index_from_id(blocks, &new_edges[0]));
+        if new_edges.is_empty() {
+            return false;
+        }
+
+        for id in &processed {
+            if let Some(i) = new_edges.iter().position(|x| x == id) {
+                new_edges.remove(i);
+            }
         }
 
+        processed.push(front);
         next.extend(new_edges);
     }
 
+    true
+}
+fn get_possible_merge_positions(
+    blocks: &[Block],
+    cf_info: &ControlFlowInfo,
+    start: Word,
+) -> Vec<usize> {
+    let mut retval = Vec::new();
+    for (idx, block) in blocks.iter().enumerate() {
+        if block_is_reverse_idom_of(blocks, cf_info, start, block.label_id().unwrap()) {
+            retval.push(idx);
```

Comment:
this runtime pains me so much to see, but yeah, prototype phase, it's okay

Reply:
you and me both